### ISSUE TITLE
Result of running python semantics with -fastKast

### ISSUE THREAD

**Issue by dwightguth** [kframework/k-legacy#13]
So it turns out my big problem was I wasn't running it on 32 bits. Which, you know, is rather embarassing. But in any event, I finally managed to run Python with the C parser, and it didn't crash. But it didn't work, either.

Here is what I got at first:

[Error] Critical: Parse error: Unexpected end of rule
    File: /home/dguth2/k/dist/include/builtins/bool.k
    Location: (23,7,23,7)

Turns out the problem here is that we are asking for rules to be a CondSentence1 and we get back a CondSentence. CondSentence1 doesn't include ensures so I am pretty sure that's wrong. So I changed it and moved forward and got this:

> > > Colon{[pos-info,area-in-file("/home/dguth2/python-semantics/python-semantics-logic.k",area(1,8,1,9,8,1))]} <<< - unimplemented yet: org.kframework.kil.loader.JavaClassesFactory
> > > Colon{[pos-info,area-in-file("/home/dguth2/python-semantics/python-semantics-logic.k",area(1,16,1,17,16,1))]} <<< - unimplemented yet: org.kframework.kil.loader.JavaClassesFactory
> > > Colon{[pos-info,area-in-file("/home/dguth2/python-semantics/python-semantics-logic.k",area(1,8,1,9,8,1))]} <<< - unimplemented yet: org.kframework.kil.loader.JavaClassesFactory
> > > Colon{[pos-info,area-in-file("/home/dguth2/python-semantics/python-semantics-logic.k",area(1,16,1,17,16,1))]} <<< - unimplemented yet: org.kframework.kil.loader.JavaClassesFactory
> > > java.lang.NullPointerException
> > >     at org.kframework.kil.visitors.BasicVisitor.visit(BasicVisitor.java:595)
> > >     at org.kframework.kil.TermCons.accept(TermCons.java:144)
> > >     at org.kframework.kil.visitors.BasicVisitor.visit(BasicVisitor.java:585)
> > >     at org.kframework.kil.Rewrite.accept(Rewrite.java:98)
> > >     at org.kframework.kil.visitors.BasicVisitor.visit(BasicVisitor.java:268)
> > >     at org.kframework.kil.visitors.BasicVisitor.visit(BasicVisitor.java:277)
> > >     at org.kframework.kil.Ambiguity.accept(Ambiguity.java:48)
> > >     at org.kframework.kil.visitors.BasicVisitor.visit(BasicVisitor.java:114)
> > >     at org.kframework.kil.Sentence.accept(Sentence.java:105)
> > >     at org.kframework.parser.generator.ParseRulesFilter.transform(ParseRulesFilter.java:97)
> > >     at org.kframework.kil.StringSentence.accept(StringSentence.java:63)
> > >     at org.kframework.kil.visitors.BasicTransformer.transform(BasicTransformer.java:57)
> > >     at org.kframework.parser.generator.ParseRulesFilter.transform(ParseRulesFilter.java:80)
> > >     at org.kframework.kil.Module.accept(Module.java:136)
> > >     at org.kframework.kil.visitors.BasicTransformer.transform(BasicTransformer.java:39)
> > >     at org.kframework.kil.Definition.accept(Definition.java:121)
> > >     at org.kframework.parser.DefinitionLoader.parseDefinition(DefinitionLoader.java:241)
> > >     at org.kframework.parser.DefinitionLoader.loadDefinition(DefinitionLoader.java:90)
> > >     at org.kframework.kompile.KompileFrontEnd.genericCompile(KompileFrontEnd.java:284)
> > >     at org.kframework.kompile.KompileFrontEnd.kompile(KompileFrontEnd.java:252)
> > >     at org.kframework.main.Main.main(Main.java:68)

TermCons is trying to turn Colon into a term when it's a terminal, so it fails. I made a fix for this too. Now I get this:

[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: ExceptClauses ::= NeList{ExceptClause,""}
   (list ( _:List ((ListItem((K:K : K2:K) ::KeyDatum) => .List)) )) _
2: Comps ::= List{Comp,""}
   (list ( _:List ((ListItem((K:K : K2:K) ::KeyDatum) => .List)) )) _
    File: /home/dguth2/python-semantics/python-semantics-common.k
    Location: (219,64,219,109)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Comps ::= List{Comp,""}
   (list ( _:List ((ListItem(K:K) => .List)) )) _
2: ExceptClauses ::= NeList{ExceptClause,""}
   (list ( _:List ((ListItem(K:K) => .List)) )) _
    File: /home/dguth2/python-semantics/python-semantics-common.k
    Location: (224,64,224,92)
    Compilation Phase: Ambiguity filter
[Error] Critical: Could not infer a unique sort for variable 'String'. Possible sorts: KLabel, Map, K
    File: /home/dguth2/python-semantics/python-semantics-common.k
    Location: (264,37,264,43)

All three of these seem to be the result of failing to prefer K syntax over user syntax. The third of these fails with an error because VariableTypeInferenceFilter is not designed to prune branches that fail from the ambiguity tree and proceed with what remains.

I also took the liberty of attempting to replace the code Radu had written with similar code that compiled under the latest version of SDF that Jurgen pointed us to on Github. We do not have a repository for the dependencies that I can commit to anymore so I can't check it in right now but if someone creates a dependencies repository on Github I can check it in, or I can email the code to anyone who requests it.

This was quite an involved task and used a decent amount of trial and error, but eventually I got something that parsed Python. This actually revealed some bugs in my definition in which I had failed to declare follow restrictions appropriately and only the fact that the old version of SDF had bugs limiting the full generality of the parse tree prevented the code from working improperly.

This eventually gave me an execution of kompile that was significantly slower than the Java parser. I haven't timed the two SglrJNI implementations side by side because they're not really comparable when they give different results, so it's possible Radu's code is fast and mine is slow for reasons we may eventually be able to debug. Or it's possible the old version of SGLR is faster. Or it's possible that they're both slow and the problem is that I have some rules with very large ambiguity trees that the Java parser handles more efficiently than the C parser. I'm not sure.

In any event, it still didn't compile, but it gave me this result:

[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: ExceptClauses ::= NeList{ExceptClause,""}
   (list ( _:List ((ListItem((K:K : K2:K) ::KeyDatum) => .List)) )) _
2: Comps ::= List{Comp,""}
   (list ( _:List ((ListItem((K:K : K2:K) ::KeyDatum) => .List)) )) _
    File: /home/dguth2/python-semantics/python-semantics-common.k
    Location: (219,64,219,109)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Comps ::= List{Comp,""}
   (list ( _:List ((ListItem(K:K) => .List)) )) _
2: ExceptClauses ::= NeList{ExceptClause,""}
   (list ( _:List ((ListItem(K:K) => .List)) )) _
    File: /home/dguth2/python-semantics/python-semantics-common.k
    Location: (224,64,224,92)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: AssignTargets ::= NeList{Target,"="}
   ([ Ts2:K ])= Ts:AssignTargets
2: AssignTargets ::= NeList{Target,"="}
   ([ Ts2:K ])= Ts:AssignTargets
    File: /home/dguth2/python-semantics/python-semantics-assignment.k
    Location: (11,52,11,78)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Stmt ::= "del" NeTargets
   del ((HOLE . _:Id), _)
2: Stmt ::= "del" NeTargets
   del ((HOLE . _:Id), _)
    File: /home/dguth2/python-semantics/python-semantics-assignment.k
    Location: (27,11,27,30)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Stmt ::= "del" NeTargets
   del ((HOLE [ _:K ]), _)
2: Stmt ::= "del" NeTargets
   del ((HOLE [ _:K ]), _)
    File: /home/dguth2/python-semantics/python-semantics-assignment.k
    Location: (28,11,28,29)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Stmt ::= "del" NeTargets
   del ((K:K [ HOLE ]), _:K)
2: Stmt ::= "del" NeTargets
   del ((K:K [ HOLE ]), _:K)
    File: /home/dguth2/python-semantics/python-semantics-assignment.k
    Location: (29,11,29,29)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Arguments ::= List{Argument,","}
   O:Object, (Id2String ( X:Id ))
2: NeExps ::= NeList{Exp,","}
   O:Object, (Id2String ( X:Id ))
    File: /home/dguth2/python-semantics/python-semantics-attribute-ref.k
    Location: (17,55,17,70)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1:
   .K => if (((((((((((((O:K is (ref ( "bool" )))) or ((O is (ref ( "NoneType" )
     )))) or ((O is (ref ( "function" ))))) or ((O is (ref ( "slice" ))))) or ((
     O is (ref ( "generator" ))))) or ((O is (ref ( "builtin-function" ))))) or
     ((O is (ref ( "builtin-method" ))))) or ((O is (ref ( "traceback" ))))) or
     ((O is (ref ( "code" ))))) or ((O is (ref ( "frame" ))))) or ((O is (ref (
     "method" ))))) or ((O is (ref ( "cell" ))))) : raiseInternal ( "TypeError"
     , "type is not an acceptable base type" ) ~> if ((((gettype ( O )) is (ref
     ( "type" )))) or (bool ( (hasbase ( Winner:K , (gettype ( O )) )) ))) :
     pass else : (test ( (hasbase ( (gettype ( O )) , Winner )) , (setWinner ( (
     gettype ( O )) )) , (raiseInternal ( "TypeError" ,
     "metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases"
      )) ))
2:
   .K => if (((((((((((((O:K is (ref ( "bool" )))) or ((O is (ref ( "NoneType" )
     )))) or ((O is (ref ( "function" ))))) or ((O is (ref ( "slice" ))))) or ((
     O is (ref ( "generator" ))))) or ((O is (ref ( "builtin-function" ))))) or
     ((O is (ref ( "builtin-method" ))))) or ((O is (ref ( "traceback" ))))) or
     ((O is (ref ( "code" ))))) or ((O is (ref ( "frame" ))))) or ((O is (ref (
     "method" ))))) or ((O is (ref ( "cell" ))))) : (raiseInternal ( "TypeError"
      , "type is not an acceptable base type" )) ~> if ((((gettype ( O )) is (
     ref ( "type" )))) or (bool ( (hasbase ( Winner:K , (gettype ( O )) )) ))) :
      pass else : (test ( (hasbase ( (gettype ( O )) , Winner )) , (setWinner (
     (gettype ( O )) )) , (raiseInternal ( "TypeError" ,
     "metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases"
      )) ))
    File: /home/dguth2/python-semantics/python-semantics-classes.k
    Location: (54,9,54,655)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1:
   "__class__" |-> #if IsMethod:Bool #then (ref ( "builtin-method" )) #else (ref
      ( "builtin-function" )) #fi
2:
   "**class**" |-> #if IsMethod:Bool #then (ref ( "builtin-method" )) #else (ref
      ( "builtin-function" )) #fi
    File: /home/dguth2/python-semantics/python-semantics-builtin-modules.k
    Location: (139,81,139,171)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1:
   "**class**" |-> #if IsMethod:K #then (ref ( "builtin-method" )) #else (ref (
     "builtin-function" )) #fi
2:
   "**class**" |-> #if IsMethod:K #then (ref ( "builtin-method" )) #else (ref (
     "builtin-function" )) #fi
    File: /home/dguth2/python-semantics/python-semantics-builtin-modules.k
    Location: (151,81,151,171)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Exp ::= Exp "+" Exp
   (tuple ( ((getref2 ( (ref ( Frame:Int )) , "f_cells" )) [ N:K ]) )) + (
     makeClosure ( (list ( L:List )) ))
2: Exp ::= Exp "+" Exp
   (tuple ( ((getref2 ( (ref ( Frame:Int )) , "f_cells" )) [ N:K ]) )) + (
     makeClosure ( (list ( L:List )) ))
    File: /home/dguth2/python-semantics/python-semantics-functions.k
    Location: (93,48,93,113)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Exps ::= List{Exp,","}
   (#if (Ex:Exp ==K (ref ( "None" ))) #then (ref ( "None" )) #else (Ex . (
     String2Id ( "**class**" ))) #fi), (Ex, (#if (Ex ==K (ref ( "None" ))) #then
      (ref ( "None" )) #else (Ex . (String2Id ( "**traceback**" ))) #fi))
2: NeExps ::= NeList{Exp,","}
   (#if (Ex:Exp ==K (ref ( "None" ))) #then (ref ( "None" )) #else (Ex . (
     String2Id ( "**class**" ))) #fi), (Ex, (#if (Ex ==K (ref ( "None" ))) #then
      (ref ( "None" )) #else (Ex . (String2Id ( "**traceback**" ))) #fi))
    File: /home/dguth2/python-semantics/python-semantics-try.k
    Location: (128,64,129,60)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1:
   invokeBuiltin ( (obj ( "weakref.**call**" , _:Bag )) , ListItem(O:Object) ,
     .Map ) => #if ((getattr ( O , "referent" )) =/=K .Obj) #then (ref ( (
     getattr ( O , "referent" )) )) #else (ref ( "None" )) #fi
2:
   invokeBuiltin ( (obj ( "weakref.__call__" , _:Bag )) , ListItem(O:Object) ,
     .Map ) => #if ((getattr ( O , "referent" )) =/=K .Obj) #then (ref ( (
     getattr ( O , "referent" )) )) #else (ref ( "None" )) #fi
    File: /home/dguth2/python-semantics/python-semantics-garbage-collection.k
    Location: (124,12,124,171)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Arguments ::= List{Argument,","}
   (ref ( "sys" )), (ref ( "_imp" ))
2: NeExps ::= NeList{Exp,","}
   (ref ( "sys" )), (ref ( "_imp" ))
    File: /home/dguth2/python-semantics/python-semantics-import.k
    Location: (15,182,15,205)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: NeExps ::= NeList{Exp,","}
   Self:K, (ref ( "None" ))
2: Arguments ::= List{Argument,","}
   Self:K, (ref ( "None" ))
    File: /home/dguth2/python-semantics/python-semantics-io.k
    Location: (77,95,77,112)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: NeExps ::= NeList{Exp,","}
   Self:K, (ref ( "None" ))
2: Arguments ::= List{Argument,","}
   Self:K, (ref ( "None" ))
    File: /home/dguth2/python-semantics/python-semantics-io.k
    Location: (85,97,85,114)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Arguments ::= List{Argument,","}
   Name:Object, (strvalue ( Mode:K ))
2: NeExps ::= NeList{Exp,","}
   Name:Object, (strvalue ( Mode:K ))
    File: /home/dguth2/python-semantics/python-semantics-io.k
    Location: (135,39,135,59)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: Arguments ::= List{Argument,","}
   (O:K . (String2Id ( "__getitem__" ))), ((ref ( "range" )) ( (\* Indices:Object
     ) ))
2: NeExps ::= NeList{Exp,","}
   (O:K . (String2Id ( "**getitem**" ))), ((ref ( "range" )) ( (\* Indices:Object
     ) ))
    File: /home/dguth2/python-semantics/python-semantics-slicing.k
    Location: (22,55,22,108)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: NeExps ::= NeList{Exp,","}
   O:Object, ((ref ( "None" )), -1)
2: Arguments ::= List{Argument,","}
   O:Object, ((ref ( "None" )), -1)
    File: /home/dguth2/python-semantics/python-semantics-strings.k
    Location: (46,87,46,105)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: NeExps ::= NeList{Exp,","}
   O:Object, (Sep:Object, -1)
2: Arguments ::= List{Argument,","}
   O:Object, (Sep:Object, -1)
    File: /home/dguth2/python-semantics/python-semantics-strings.k
    Location: (47,108,47,118)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: NeExps ::= NeList{Exp,","}
   O:Object, (substrString ( FieldName:K , 1 , (findCharsInString ( FieldName ,
     ".[" , 1 )) ))
2: Arguments ::= List{Argument,","}
   O:Object, (substrString ( FieldName:K , 1 , (findCharsInString ( FieldName ,
     ".[" , 1 )) ))
    File: /home/dguth2/python-semantics/python-semantics-strings.k
    Location: (94,207,94,275)
    Compilation Phase: Ambiguity filter
[Warning] Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: NeExps ::= NeList{Exp,","}
   (convertField ( O:Object , (substrString ( Field:K , (findCharsInString (
     Field , ":!" , 0 )) , ((findCharsInString ( Field , ":!" , 0 )) +Int 3) ))
     )), (substrString ( Field , ((findCharsInString ( Field , ":" , 0 )) +Int 1
     ) , (lengthString ( Field )) ))
2: Arguments ::= List{Argument,","}
   (convertField ( O:Object , (substrString ( Field:K , (findCharsInString (
     Field , ":!" , 0 )) , ((findCharsInString ( Field , ":!" , 0 )) +Int 3) ))
     )), (substrString ( Field , ((findCharsInString ( Field , ":" , 0 )) +Int 1
     ) , (lengthString ( Field )) ))
    File: /home/dguth2/python-semantics/python-semantics-strings.k
    Location: (108,79,108,275)
    Compilation Phase: Ambiguity filter
[Error] Compiler: Unbounded Variable Frame:K .
    File: /home/dguth2/python-semantics/python-semantics-garbage-collection.k
    Location: (80,67,80,72)
    Compilation Phase: class org.kframework.compile.checks.CheckVariables
[Error] Compiler: Unbounded Variable C:Bag .
    File: /home/dguth2/python-semantics/python-semantics-garbage-collection.k
    Location: (70,217,70,218)
    Compilation Phase: class org.kframework.compile.checks.CheckVariables
[Error] Compiler: Unbounded Variable K:K .
    File: /home/dguth2/python-semantics/python-semantics-garbage-collection.k
    Location: (58,88,58,89)
    Compilation Phase: class org.kframework.compile.checks.CheckVariables

This includes some of the errors that I had before but also includes new errors that seem to be the result of AmbDuplicateFilter and TypeInferenceSupremumFilter not working correctly on the new kil classes. Why that may be I didn't investigate. The final errors I collected from multiple nondeterministic runs of the command (and I may not have all of them, either) and seem to result from certain variable terms not showing up in the list of variables found in the right-hand side of a rule. Again, why that is I don't know and I haven't investigated.

The good news out of all of this is that while there is still plenty of work left to be done, at no point did the C parser crash or fail to produce the correct data. Which means that all the work that really remains is to fix up the K framework to work with all the nuances and details.

Anyway, this concludes my expedition into the C parser. If you want any more research to be done on this, you will have to get either Radu or whoever will inherit my Python semantics to do it.

**Comment by grosu** [comment]
Thanks, Dwight.  How about performance?

Grigore

Sent from phone

-------- Original message --------

**Comment by dwightguth** [comment]
As I said the C parser was actually slower for me, about 150 seconds instead of 80. This I think is because some of the rules that have very big associative lists took a very long time to parse.

**Comment by radumereuta** [comment]
Thanks Dwight.

This actually helps. I remember that I have to fix some other tricks that I implemented in the grammars in order to reduce the number of ambiguities.

**Comment by grosu** [comment]
(Dwight, yesterday I answered from my phone, on the road, so I missed the sentence saying the C parser was slower; but it is good that you sent us the numbers, too)

150s vs 80s: It would be interesting to also look how long it took to parse each rule, as Radu used to do.  Maybe it is only one or two rules which are slow.

Radu, what is the status of the parser now?  Monday you said that you were out of ideas, so you needed help.  Stefan found that simple bug, so things unblocked.  Dwight further gave you lots of errors, even fixed some things.  What next?  What is currently holding us from declaring victory and switching to the new, C parser?

Stefan, Radu, and others: how should we deal with the 64-bit issue?  Could it be possible to contact the guy that Jurgen mentioned inhis message, who actually produced the 64-bit version?  Maybe he remembers what happened and how to make it work?

Grigore

**Comment by radumereuta** [comment]
The next phase is a bit tricky. Dwight managed to stumble upon the last few things that I didn't translate into java, that is some piece of Stratego code that would handle some rare bugs in SDF for which I only have some workaround: http://yellowgrass.org/issue/Spoofax/394 - actually what is on that page is only half of it. I had to modify that code multiple times until I got to a stable version. Some other cosmetic stuff, and it should work.

Now, regarding the timing issues. Dwight, if you run kompile with the option -v, it will create a timing.log file in the -kompiled directory with the time it takes to parse and disambiguate each rule. I've stumbled upon that problem myself. If you remember I've told you about BF and the fact that the parser wouldn't finish even after 18 minutes, when I got bored and closed it.

I'll have to look over the timing information to see what exactly happens, but I think we are seeing that the C parser has it's limitations too.
